File: src/main/java/net/casualuhc/uhcmod/utils/screen/CustomScreen.java
package net.casualuhc.uhcmod.utils.screen;

import net.casualuhc.uhcmod.utils.Scheduler;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.inventory.Inventory;
import net.minecraft.inventory.SimpleInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.screen.ScreenHandler;
import net.minecraft.screen.ScreenHandlerType;
import net.minecraft.screen.slot.Slot;
import net.minecraft.screen.slot.SlotActionType;

import java.util.function.Consumer;

public abstract class CustomScreen extends ScreenHandler {;
	private final PlayerInventory playerInventory;
	private final Inventory inventory;

	/**
	 * Creates a custom screen with a given player inventory.
	 */
	public CustomScreen(PlayerInventory playerInventory, int syncId, int rows) {
		super(rowsToType(rows), syncId);
		this.playerInventory = playerInventory;
		this.inventory = new SimpleInventory(9 * rows);

		int i = (rows - 4) * 18;

		for (int j = 0; j < rows; j++) {
			for (int k = 0; k < 9; k++) {
				this.addSlot(new Slot(this.inventory, k + j * 9, 8 + k * 18, 18 + j * 18));
			}
		}

		for (int j = 0; j < 3; j++) {
			for (int k = 0; k < 9; k++) {
				this.addSlot(new Slot(playerInventory, k + j * 9 + 9, 8 + k * 18, 103 + j * 18 + i));
			}
		}

		for (int j = 0; j < 9; j++) {
			this.addSlot(new Slot(playerInventory, j, 8 + j * 18, 161 + i));
		}
	}

	/**
	 * Creates a custom screen with an empty player inventory.
	 */
	public CustomScreen(PlayerEntity player, int syncId, int rows) {
		this(new PlayerInventory(player), syncId, rows);
	}

	public void modifyInventory(Consumer<Inventory> inventoryModifier) {
		inventoryModifier.accept(this.inventory);
	}

	public void modifyPlayerInventory(Consumer<Inventory> inventoryModifier) {
		inventoryModifier.accept(this.playerInventory);
	}


	@Override
	public abstract void onSlotClick(int slotIndex, int button, SlotActionType actionType, PlayerEntity player);

	@Override
	public final ItemStack transferSlot(PlayerEntity player, int index) {
		throw new UnsupportedOperationException("Cannot transfer slots from within a CustomScreen");
	}

	@Override
	public boolean canUse(PlayerEntity player) {
		return true;
	}

	@Override
	public void close(PlayerEntity player) {
		super.close(player);
		player.playerScreenHandler.syncState();
		// Scheduler.schedule(0, player.playerScreenHandler::syncState);
	}

	private static ScreenHandlerType<?> rowsToType(int rows) {
		return switch (rows) {
			case 1 -> ScreenHandlerType.GENERIC_9X1;
			case 2 -> ScreenHandlerType.GENERIC_9X2;
			case 3 -> ScreenHandlerType.GENERIC_9X3;
			case 4 -> ScreenHandlerType.GENERIC_9X4;
			case 5 -> ScreenHandlerType.GENERIC_9X5;
			case 6 -> ScreenHandlerType.GENERIC_9X6;
			default -> throw new IllegalStateException("Invalid number of rows: " + rows);
		};
	}
}

File: src/main/kotlin/net/casualuhc/uhcmod/screen/CustomScreen.kt
package net.casualuhc.uhcmod.screen

import net.minecraft.world.Container
import net.minecraft.world.SimpleContainer
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.inventory.ClickType
import net.minecraft.world.inventory.MenuType
import net.minecraft.world.inventory.Slot
import net.minecraft.world.item.ItemStack
import java.util.function.Consumer

abstract class CustomScreen(
    playerInventory: Inventory,
    syncId: Int,
    rows: Int
): AbstractContainerMenu(rowsToType(rows), syncId) {
    private val playerInventory: Inventory
    private val inventory: Container

    init {
        this.playerInventory = playerInventory
        this.inventory = SimpleContainer(9 * rows)
        val i = (rows - 4) * 18
        for (j in 0 until rows) {
            for (k in 0..8) {
                this.addSlot(Slot(this.inventory, k + j * 9, 8 + k * 18, 18 + j * 18))
            }
        }
        for (j in 0..2) {
            for (k in 0..8) {
                this.addSlot(Slot(playerInventory, k + j * 9 + 9, 8 + k * 18, 103 + j * 18 + i))
            }
        }
        for (j in 0..8) {
            this.addSlot(Slot(playerInventory, j, 8 + j * 18, 161 + i))
        }
    }

    constructor(player: Player, syncId: Int, rows: Int): this(Inventory(player), syncId, rows)

    fun modifyInventory(inventoryModifier: Consumer<Container>) {
        inventoryModifier.accept(this.inventory)
    }

    fun modifyPlayerInventory(inventoryModifier: Consumer<Inventory>) {
        inventoryModifier.accept(this.playerInventory)
    }

    override fun clicked(slotId: Int, button: Int, clickType: ClickType, player: Player) {
        super.clicked(slotId, button, clickType, player)
    }

    override fun quickMoveStack(player: Player, slot: Int): ItemStack {
        throw UnsupportedOperationException("Cannot transfer slots from within a CustomScreen")
    }

    override fun stillValid(player: Player): Boolean {
        return true
    }

    override fun removed(player: Player) {
        super.removed(player)
        player.containerMenu.sendAllDataToRemote()
        // Scheduler.schedule(0, player.playerScreenHandler::syncState);
    }

    companion object {
        private fun rowsToType(rows: Int): MenuType<*> {
            return when (rows) {
                1 -> MenuType.GENERIC_9x1
                2 -> MenuType.GENERIC_9x2
                3 -> MenuType.GENERIC_9x3
                4 -> MenuType.GENERIC_9x4
                5 -> MenuType.GENERIC_9x5
                6 -> MenuType.GENERIC_9x6
                else -> throw IllegalStateException("Invalid number of rows: $rows")
            }
        }
    }
}
